add option for user to upload manifest in Subscriptions tab

isratrade opened the pull request.

isratrade commented:
Don't merge until @jmrodri PR for backend functionality to save the manifest file and request a GET api endpoint for the review subscriptions tab.

isratrade commented on fusor-ember-cli/app/controllers/subscriptions/credentials.js:
```diff
@@ -53,7 +57,62 @@ export default Ember.Controller.extend({
+      //ic-ajax request
+      request({
+        type: 'POST',
+        url: '/fusor/api/v21/subscriptions/upload',
```
@jmrodri, I'm using this endpoint that you mentioned in IRC and using POST. I can change to PUT if you want.

isratrade commented on fusor-ember-cli/app/controllers/subscriptions/credentials.js:
```diff
@@ -53,7 +57,62 @@ export default Ember.Controller.extend({
+      var formData = new FormData();
+      formData.append('manifest_file[name]', manifestFile.name);
+      formData.append('manifest_file[file]', manifestFile);
+      formData.append('manifest_file[deployment_id]', this.get('deploymentId'));
```
content-type is 'multipart/form-data', not json
I'm passing `deployment_id` as what of the payload

jmrodri replied:
Excellent. I'll update the endpoint to look for deployment id. I thought about that on my way to work this morning that I need to know the deployment so I can associate the two.

isratrade commented on fusor-ember-cli/app/templates/subscriptions/credentials.hbs:
```diff
@@ -43,36 +28,119 @@
+<div class='row'>
+  <div class='col-md-9'>
+    <p>
+      Select a content provider for this RHCI deployment.
```
waiting on sj and mreid for update text here.

isratrade commented on fusor-ember-cli/app/templates/subscriptions/review-subscriptions.hbs:
```diff
@@ -0,0 +1,42 @@
+        <tr>
+          <td colspan="8">
+            <p class='no_subscriptions'>
+              <h3> Get JSON response for API endpoint that parses manifests and display table here</h3>
```
@waiting on @jmrodri API and what to show for the `review subscriptions` page. Will it be the same columns as on the `select subscriptions` page, just with no checkbox??

jmrodri replied:
I haven't determined what that's going to return yet. Depends on what I can glean from the manifest.

isratrade commented on server/app/controllers/fusor/api/v21/subscriptions_controller.rb:
```diff
@@ -13,6 +13,18 @@
+    def check_content_type
+      # overwrite this method in V2::BaseController that
+      # renders error if request.content_type != "application/json"
+      # the upload method uses "multipart/form-data"
```
@jmrodri, this is needed to overwrite `check_content_type` so Sat6 doesn't give an error for wrong MIME type

jmrodri replied:
@daviddavis & I decided to use
```ruby
skip_before_filter :check_content_type, :only => [:upload]
```

That way we don't have to override the check_content_type method.

isratrade replied:
@jmrodri, we are overriding `check_content_type` for the `upload` method only. otherwise, it gives an error for not being json mime type.

daviddavis replied:
@isratrade skipping the before filter should have the same effect and it's more concise IMO.

isratrade commented on server/app/controllers/fusor/api/v21/subscriptions_controller.rb:
```diff
@@ -13,6 +13,19 @@
+
+    def upload
+      # placeholder only. TODO perform actually upload file
+      render :json => {params: params, request: @request}
```
@jmrodri, here's a placeholder route and method for `POST fusor/api/v21/subscriptions/upload` and you will be see in the "network" console of your browser that you have all the info that you need.

jmrodri replied:
I have that already, except my render is different. I'll update my end to match. I just had render :json => {}

isratrade replied:
@jmrodri, right, returning `:json => {}` is great. The other code was just placeholder so I got see better that everything was there.

jmrodri replied:
@daviddavis & I decided to use
```ruby
skip_before_filter :check_content_type, :only => [:upload]
```

That way we don't have to override the check_content_type method.

jmrodri commented:
I tried applying this as a patch to my disconnected branch and it failed miserably. I tried doing it as follows:

```
wget https://github.com/fusor/fusor/pull/415.patch
git am 415.patch
```

That failed with errors. Then I tried to do it as follows:

```
git apply .git/rebase-apply/0001 --reject
```
this updated a bunch of files, and reject a few. I then updated the rejected files by hand. But it no longers gives you credit for the work which is not fair.

The ```.js``` and ```.hbs``` changes are fine. Remove the changes to ```subscriptions_controller.rb``` and ```v21.rb```, they're already done in my PR #426.

jmrodri commented:
I figured out how to do this by hand. So I will do it tomorrow morning and clean this PR up.
http://www.pizzhacks.com/bugdrome/2011/10/deal-with-git-am-failures/

isratrade commented:
@jmrodri, I uploaded this PR so now it should not have merge conflicts with your PR https://github.com/fusor/fusor/pull/426

* I removed my changes to `server/app/controllers/fusor/api/v21/subscriptions_controller.rb` and `server/config/routes/api/v21.rb` since you already have them.
* I changed from POST to PUT 'upload', as this is what you defined in route.
* I updated the text per the current wireframes

isratrade commented on fusor-ember-cli/app/templates/subscriptions/credentials.hbs:
```diff
@@ -43,36 +28,129 @@
+  <div class='col-md-9'>
+    <p>
+      Satellite will have external network access and receive content directly from Red Hat Customer Portal.
+    </p>
```
Here's some of the new wording from the wireframes

isratrade commented on fusor-ember-cli/app/templates/subscriptions/credentials.hbs:
```diff
@@ -43,36 +28,129 @@
+          <p>
+            Satellite needs a URL within the network from which it can access content.
+          </p>
+          {{text-f label="Content Mirror URL" value=cdnUrl disabled=false}}
```
and this label 'Content Mirror URL' instead of 'Base Manifest URL'

isratrade commented on fusor-ember-cli/app/templates/subscriptions/credentials.hbs:
```diff
@@ -43,36 +28,129 @@
+          <p>
+            <br />
+            Subscriptions that are managed by Satellite need to be added through a subscription management application manifest, which you create and obtain through Customer Portal. This manifest will be assigned to an organization, and any system that registers with that organization can use those subscriptions.
+          </p>
```
word new wording above per the wireframes.
